File: market/test/java/ru/yandex/market/wms/timetracker/controller/AssignTaskControllerTest.java
package ru.yandex.market.wms.timetracker.controller;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.junit.jupiter.api.Test;
import org.mockito.ArgumentMatchers;
import org.mockito.Mockito;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.context.annotation.Import;
import org.springframework.http.MediaType;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders;
import org.springframework.test.web.servlet.result.MockMvcResultMatchers;

import ru.yandex.market.wms.timetracker.config.NoAuthTest;
import ru.yandex.market.wms.timetracker.dto.AssignTaskRequest;
import ru.yandex.market.wms.timetracker.model.enums.EmployeeStatus;
import ru.yandex.market.wms.timetracker.service.AssignTaskService;

@WebMvcTest(AssignTaskController.class)
@ActiveProfiles("test")
@Import(NoAuthTest.class)
class AssignTaskControllerTest {

    @MockBean
    private AssignTaskService assignTaskService;

    @Autowired
    private ObjectMapper mapper;

    @Autowired
    private MockMvc mockMvc;

    @Test
    void assignTask() throws Exception {

        AssignTaskRequest contentExpected = AssignTaskRequest.builder()
                .assigner("assigner")
                .duration(15L)
                .status(EmployeeStatus.SHIPPING)
                .userNames(List.of("test1", "test2"))
                .build();

        Mockito.doNothing().when(assignTaskService).save(
                ArgumentMatchers.eq("SOF"), ArgumentMatchers.any(AssignTaskRequest.class));

        mockMvc
                .perform(
                        MockMvcRequestBuilders.post("/assign-task/SOF")
                                .contentType(MediaType.APPLICATION_JSON_VALUE)
                                .accept(MediaType.APPLICATION_JSON)
                                .content(mapper.writeValueAsString(contentExpected)))
                .andExpect(MockMvcResultMatchers.status().is2xxSuccessful());
    }

    @Test
    void assignTaskWhenOnlyOneUser() throws Exception {

        AssignTaskRequest contentExpected = AssignTaskRequest.builder()
                .assigner("assigner")
                .duration(15L)
                .status(EmployeeStatus.SHIPPING)
                .userNames(List.of("test1"))
                .build();

        Mockito.doNothing().when(assignTaskService).save(
                ArgumentMatchers.eq("SOF"), ArgumentMatchers.any(AssignTaskRequest.class));

        mockMvc
                .perform(
                        MockMvcRequestBuilders.post("/assign-task/SOF")
                                .contentType(MediaType.APPLICATION_JSON_VALUE)
                                .accept(MediaType.APPLICATION_JSON)
                                .content(mapper.writeValueAsString(contentExpected)))
                .andExpect(MockMvcResultMatchers.status().is2xxSuccessful());
    }

    @Test
    void assignTaskWhenDurationIsEmpty() throws Exception {

        AssignTaskRequest contentExpected = AssignTaskRequest.builder()
                .assigner("assigner")
                .duration(0L)
                .status(EmployeeStatus.SHIPPING)
                .userNames(List.of("test1"))
                .build();

        mockMvc
                .perform(
                        MockMvcRequestBuilders.post("/assign-task/SOF")
                                .contentType(MediaType.APPLICATION_JSON_VALUE)
                                .accept(MediaType.APPLICATION_JSON)
                                .content(mapper.writeValueAsString(contentExpected)))
                .andExpect(MockMvcResultMatchers.status().is4xxClientError());
    }

    @Test
    void assignTaskWhenAssignerIsEmpty() throws Exception {

        AssignTaskRequest contentExpected = AssignTaskRequest.builder()
                .assigner("")
                .duration(15L)
                .status(EmployeeStatus.SHIPPING)
                .userNames(List.of("test1"))
                .build();

        mockMvc
                .perform(
                        MockMvcRequestBuilders.post("/assign-task/SOF")
                                .contentType(MediaType.APPLICATION_JSON_VALUE)
                                .accept(MediaType.APPLICATION_JSON)
                                .content(mapper.writeValueAsString(contentExpected)))
                .andExpect(MockMvcResultMatchers.status().is4xxClientError());
    }

    @Test
    void assignTaskWhenUserNamesIsEmpty() throws Exception {

        AssignTaskRequest contentExpected = AssignTaskRequest.builder()
                .assigner("assigner")
                .duration(15L)
                .status(EmployeeStatus.SHIPPING)
                .userNames(Collections.emptyList())
                .build();

        mockMvc
                .perform(
                        MockMvcRequestBuilders.post("/assign-task/SOF")
                                .contentType(MediaType.APPLICATION_JSON_VALUE)
                                .accept(MediaType.APPLICATION_JSON)
                                .content(mapper.writeValueAsString(contentExpected)))
                .andExpect(MockMvcResultMatchers.status().is4xxClientError());
    }

    @Test
    void assignTaskWhenUserNamesContainsNull() throws Exception {

        AssignTaskRequest contentExpected = AssignTaskRequest.builder()
                .assigner("assigner")
                .duration(15L)
                .status(EmployeeStatus.SHIPPING)
                .userNames(Arrays.asList(null, "test"))
                .build();

        mockMvc
                .perform(
                        MockMvcRequestBuilders.post("/assign-task/SOF")
                                .contentType(MediaType.APPLICATION_JSON_VALUE)
                                .accept(MediaType.APPLICATION_JSON)
                                .content(mapper.writeValueAsString(contentExpected)))
                .andExpect(MockMvcResultMatchers.status().is4xxClientError());
    }

    @Test
    void assignTaskWhenUserNamesContainsEmptyString() throws Exception {

        AssignTaskRequest contentExpected = AssignTaskRequest.builder()
                .assigner("assigner")
                .duration(15L)
                .status(EmployeeStatus.SHIPPING)
                .userNames(Arrays.asList("", "test"))
                .build();

        mockMvc
                .perform(
                        MockMvcRequestBuilders.post("/assign-task/SOF")
                                .contentType(MediaType.APPLICATION_JSON_VALUE)
                                .accept(MediaType.APPLICATION_JSON)
                                .content(mapper.writeValueAsString(contentExpected)))
                .andExpect(MockMvcResultMatchers.status().is4xxClientError());
    }
}

File: market/test/kotlin/ru/yandex/market/wms/taskrouter/task/controller/AssignTaskControllerTest.kt
package ru.yandex.market.wms.taskrouter.task.controller

import com.github.springtestdbunit.annotation.DatabaseSetup
import com.github.springtestdbunit.annotation.DatabaseSetups
import com.github.springtestdbunit.annotation.ExpectedDatabase
import com.github.springtestdbunit.assertion.DatabaseAssertionMode
import org.junit.jupiter.api.Test
import org.springframework.http.MediaType
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders
import org.springframework.test.web.servlet.result.MockMvcResultMatchers
import ru.yandex.market.wms.taskrouter.config.BaseTest

internal class AssignTaskControllerTest : BaseTest(){

    @Test
    @DatabaseSetups(
        DatabaseSetup("/taskmanagement/db/common.xml")
    )
    @ExpectedDatabase(
        value = "/taskmanagement/db/assigntask/after.xml",
        assertionMode = DatabaseAssertionMode.NON_STRICT_UNORDERED
    )
    fun assignUserToZone() {
        mockMvc.perform(
            MockMvcRequestBuilders.post("/task/assign")
                .contentType(MediaType.APPLICATION_JSON_VALUE)
                .accept(MediaType.APPLICATION_JSON)
                .content(
                    """
                    {
                        "user":"some_user",
                        "zone":"MEZONIN_2",
                        "process":"PICKING"
                    }
                    """
                )
        )
            .andExpect(MockMvcResultMatchers.status().isOk)
    }

    @Test
    @DatabaseSetups(
        DatabaseSetup("/taskmanagement/db/common.xml")
    )
    @ExpectedDatabase(
        value = "/taskmanagement/db/assigntask/after-expected-end-time.xml",
        assertionMode = DatabaseAssertionMode.NON_STRICT_UNORDERED
    )
    fun assignUserToZoneWithExpectedEndTime() {
        mockMvc.perform(
            MockMvcRequestBuilders.post("/task/assign")
                .contentType(MediaType.APPLICATION_JSON_VALUE)
                .accept(MediaType.APPLICATION_JSON)
                .content(
                    """
                    {
                        "user":"some_user",
                        "zone":"MEZONIN_2",
                        "process":"PICKING",
                        "expectedEndTime": "2022-04-25T15:00:00"
                    }
                    """
                )
        )
            .andExpect(MockMvcResultMatchers.status().isOk)
    }

    @Test
    @DatabaseSetups(
        DatabaseSetup("/taskmanagement/db/common.xml")
    )
    fun assignUserToInvalidZone() {
        mockMvc.perform(
            MockMvcRequestBuilders.post("/task/assign")
                .contentType(MediaType.APPLICATION_JSON_VALUE)
                .accept(MediaType.APPLICATION_JSON)
                .content(
                    """
                    {
                        "user":"some_user",
                        "zone":"NOT_EXISTS",
                        "process":"PICKING"
                    }
                    """
                )
        )
            .andExpect(MockMvcResultMatchers.status().is4xxClientError)
    }

    @Test
    @DatabaseSetups(
        DatabaseSetup("/taskmanagement/db/common.xml")
    )
    @ExpectedDatabase(
        value = "/taskmanagement/db/assigntask/after.xml",
        assertionMode = DatabaseAssertionMode.NON_STRICT_UNORDERED
    )
    fun assignUserToZoneTwice() {
        mockMvc.perform(
            MockMvcRequestBuilders.post("/task/assign")
                .contentType(MediaType.APPLICATION_JSON_VALUE)
                .accept(MediaType.APPLICATION_JSON)
                .content(
                    """
                    {
                        "user":"some_user",
                        "zone":"MEZONIN_1",
                        "process":"PICKING"
                    }
                    """
                )
        )
            .andExpect(MockMvcResultMatchers.status().isOk)

        mockMvc.perform(
            MockMvcRequestBuilders.post("/task/assign")
                .contentType(MediaType.APPLICATION_JSON_VALUE)
                .accept(MediaType.APPLICATION_JSON)
                .content(
                    """
                    {
                        "user":"some_user",
                        "zone":"MEZONIN_2",
                        "process":"PICKING"
                    }
                    """
                )
        )
            .andExpect(MockMvcResultMatchers.status().isOk)
    }

    @Test
    @DatabaseSetups(
        DatabaseSetup("/taskmanagement/db/common.xml"),
        DatabaseSetup("/taskmanagement/db/assigntask-remove/before.xml")
    )
    @ExpectedDatabase(
        value = "/taskmanagement/db/assigntask-remove/after.xml",
        assertionMode = DatabaseAssertionMode.NON_STRICT_UNORDERED
    )
    fun removeAssignZone() {
        mockMvc.perform(
            MockMvcRequestBuilders.post("/task/assign/remove")
                .contentType(MediaType.APPLICATION_JSON_VALUE)
                .accept(MediaType.APPLICATION_JSON)
                .content(
                    """
                    {
                        "expectedEndTime":"2022-04-25T15:00:00"
                    }
                    """
                )
        )
            .andExpect(MockMvcResultMatchers.status().isOk)
    }

    @Test
    @DatabaseSetups(
        DatabaseSetup("/taskmanagement/db/common.xml"),
        DatabaseSetup("/taskmanagement/db/assigntask-remove/before.xml")
    )
    @ExpectedDatabase(
        value = "/taskmanagement/db/assigntask-remove/after-time-user.xml",
        assertionMode = DatabaseAssertionMode.NON_STRICT_UNORDERED
    )
    fun removeAssignZoneTimeAndUser() {
        mockMvc.perform(
            MockMvcRequestBuilders.post("/task/assign/remove")
                .contentType(MediaType.APPLICATION_JSON_VALUE)
                .accept(MediaType.APPLICATION_JSON)
                .content(
                    """
                    {
                        "user": "user1",
                        "expectedEndTime":"2022-04-25T15:00:00"
                    }
                    """
                )
        )
            .andExpect(MockMvcResultMatchers.status().isOk)
    }

    @Test
    @DatabaseSetups(
        DatabaseSetup("/taskmanagement/db/common.xml"),
        DatabaseSetup("/taskmanagement/db/assigntask-remove/before.xml")
    )
    @ExpectedDatabase(
        value = "/taskmanagement/db/assigntask-remove/after-user.xml",
        assertionMode = DatabaseAssertionMode.NON_STRICT_UNORDERED
    )
    fun removeAssignZoneOnlyUser() {
        mockMvc.perform(
            MockMvcRequestBuilders.post("/task/assign/remove")
                .contentType(MediaType.APPLICATION_JSON_VALUE)
                .accept(MediaType.APPLICATION_JSON)
                .content(
                    """
                    {
                        "user": "user3"
                    }
                    """
                )
        )
            .andExpect(MockMvcResultMatchers.status().isOk)
    }
}
